### PR TITLE
Handle PrimitiveResult as result type.

### LINKED ISSUE #334: Job results from IBMQ devices are not (always) `SamplerResult`
The code currently assumes they are and errors out if not: https://github.com/CQCL/pytket-qiskit/blob/1d027ebc5c08e6aaa5265b0c7f2c6920c73f1187/pytket/extensions/qiskit/backends/ibm.py#L617

The actual type is `PrimitiveResult`.

(This issue does not affect the latest released version, but we should fix before the next release.)

### PR DESCRIPTION
**Pull request by cqc-alec**
# Description

We were assuming a `SamplerResult`. We still handle that case, though I don't know if it ever occurs.

# Related issues

Fixes #334 .

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented hard-to-understand parts of my code.
- [ ] I have made corresponding changes to the public API documentation.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: pytket/extensions/qiskit/backends/ibm.py [hunk 1]
```diff
@@ -614,11 +622,19 @@ def get_result(self, handle: ResultHandle, **kwargs: KwargTypes) -> BackendResul
                         sleep(10)
 
                 res = job.result(timeout=kwargs.get("timeout", None))
-            assert isinstance(res, SamplerResult)
-            for circ_index, (r, d) in enumerate(zip(res.quasi_dists, res.metadata)):
-                self._ibm_res_cache[(jobid, circ_index)] = Counter(
-                    {n: int(0.5 + d["shots"] * p) for n, p in r.items()}
-                )
+            if isinstance(res, SamplerResult):
+                # TODO Is this code still reachable?
```
Comment (sjdilkes) [inline pytket/extensions/qiskit/backends/ibm.py:626]: Do we need to address this TODO here?

**Review by sjdilkes — CHANGES_REQUESTED** [review]
LGTM, just want to check the TODO doesn't need to be addressed.

File: pytket/extensions/qiskit/backends/ibm.py (hunk 1 above)
Comment (cqc-alec) [reply pytket/extensions/qiskit/backends/ibm.py:626]: I think it can wait; I will raise a new issue.

Comment (cqc-alec) [reply pytket/extensions/qiskit/backends/ibm.py:626]: #336

**Review by sjdilkes — APPROVED** [review]

File: pytket/extensions/qiskit/backends/ibm.py [hunk 2]
```diff
@@ -614,11 +622,19 @@ def get_result(self, handle: ResultHandle, **kwargs: KwargTypes) -> BackendResul
...
+                for circ_index, (r, d) in enumerate(zip(res.quasi_dists, res.metadata)):
+                    self._ibm_res_cache[(jobid, circ_index)] = Counter(
+                        {n: int(0.5 + d["shots"] * p) for n, p in r.items()}
+                    )
+            else:
+                assert isinstance(res, PrimitiveResult)
+                for circ_index, pub_result in enumerate(res._pub_results):
+                    readouts = pub_result.data.c.array
```
Comment (mwpb) [inline pytket/extensions/qiskit/backends/ibm.py:634]: Is there something in the set up that allows us to assume that the register is called `c`?

Comment (mwpb) [reply pytket/extensions/qiskit/backends/ibm.py:634]: For me it is called `readout` for instance.

Comment (mwpb) [reply pytket/extensions/qiskit/backends/ibm.py:634]: A quick question: does it make sense to have multiple classical registers in a BackendResult? Or would we end up 'concatenating' them into one? If so then something like the following might prove useful?
```
counts = []
for result in res._pub_results:
    result = SamplerPubResult(result.data)
    counts.append(result.join_data().get_counts())
```
It gives a dictionary of the form `{'00': 7, '11': 3}` for instance. Hopefully that is reasonably easy to convert into a `BackendResult`? Alternatively I suppose we could use the shots directly but that would be a lot large object I think.

Comment (cqc-alec) [reply pytket/extensions/qiskit/backends/ibm.py:634]: Yes it does make sense, we just pack them in lexicographic order I think. Thanks for the suggestion -- I'll try it out.

Comment (mwpb) [reply pytket/extensions/qiskit/backends/ibm.py:634]: (Note that I've also ended up with a list of BackendResults, which I'm not quite sure if I should be merging or not. I think they can correspond to completely different circuits.)

**Pull request closed without merging**